docs/criteria.md

```diff
@@ -0,0 +1,877 @@
+
+`AbstractCriteria`, in turn, provides a general methods to work with criteria regardless of their meaning.
+
+## Criterion
+
+Criterion is a user selectable predefined condition.
+
+**Why selectable?** Because admins/users can select them (remember trophy creation process).
+
+**Why predefined?** Because XenForo already knows how to handle them (using criteria classes methods).
+
+Every criterion consists of two parts: **rule** and (optionally) **data**.
+
+### Rule
+
+The criterion rule is simply a sting in [snake case](https://en.wikipedia.org/wiki/Snake_case) (words_are_separated_with_underscore_character).
+
+It has two essential purposes:
+
+1. It is used to distinguish criteria
+2. When performing matching, the rule is converted into a [camel case](https://en.wikipedia.org/wiki/Camel_case) name of a method that handles this criterion (see ["How criteria works"](#how-criteria-works)).
+
+### Data
+
+It is just an optional array of additional criterion data. For example, "User has posted at least X messages" criterion has a data array with one element: a number of messages.
+
+## How criteria system works
+
+In this sections, we describe how criteria system works from A to Z.
+
+### Template
+
+It all starts from template code. Here is how criteria look inside templates:
+
+```html
+<xf:checkbox label="Criteria container">
+    
+    <!-- Criterion -->
+    <xf:option name="foo_criteria[criterion_1_rule][rule]" value="criterion_1_rule" ... />
+    
+    <!-- Criterion with data -->
+    <xf:option name="bar_criteria[criterion_2_rule][rule]" value="criterion_2_rule" ... >
+        <xf:... name="bar_criteria[criterion_2_rule][data][param_1]" ... />
+        <xf:... name="bar_criteria[criterion_2_rule][data][param_2]" ... />
+    </xf:option>
+        
+</xf:checkbox>
+```
+
+As you can see, criterion is simply a checkbox with optional input fields inside (criterion data). Let's analyze the code:
+
+-   `foo_criteria` and `bar_criteria` are the input containers and usually `foo` and `bar` parts refer to criteria type. For example, `user_criteria[...]` lets us know that this criteria belong to User criteria.
+-   `value="criterion_1_rule"` and `value="criterion_2_rule"` are, obviously, the rules of criteria.
+
+!!! note
+    Keep in mind that `criterion_1/2_rule` in `name` attributes may not have to be criteria rules! These are just names for input containers. You can easily write `<xf:option name="foo[bar][rule]" value="criterion_rule" />` and it will work correctly. The criterion rule will be `criterion_rule`, not `bar`.
+
+### (Optionally) Storing selected criteria
+
+Inside the controller, the criteria form data from the previous section can be filtered, serialized and saved in database columns of `mediumblob` type for better days:
+
+```php
+$fooCriteriaInput = $this->filter('foo_criteria', 'array');
+$barCriteriaInput = $this->filter('bar_criteria', 'array');
+
+$form->basicEntitySave($bazEntity, [$fooCriteriaInput, $barCriteriaInput]);
```

Comment:
Looking at this line of code, I'm not sure it works as expected. Would this be better as:

```
$form->basicEntitySave($bazEntity, [
	'foo_criteria' => $fooCriteriaInput,
	'bar_criteria' => $barCriteriaInput
]);